File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.java
class Solution{
    public int numberOfSubstrings(String s){
        int n=s.length();

        HashMap<Character, Integer> mp=new HashMap<>();
        int res=0;
        int i=0, j=0;
        while(j<n){
            mp.put(s.charAt(j), mp.getOrDefault(s.charAt(j), 0) + 1);
            while(i<=j && mp.size() == 3){
                res += (n-j);
                char ch=s.charAt(i);
                mp.put(ch, mp.get(ch)-1);

                if(mp.get(ch) == 0){
                    mp.remove(ch);
                }

                i++;
            }
            j++;
        }

        return res;

    }
}

File: 1460-number-of-substrings-containing-all-three-characters/number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n=len(s)

        mp={}
        res=0
        i, j = 0, 0
        while j<n:
            mp[s[j]]=mp.get(s[j], 0)+1
            while len(mp) ==  3:
                res += (n-j)
                mp[s[i]] -= 1

                if mp[s[i]] == 0:
                    del mp[s[i]]

                i+=1
            j+=1

        return res